// @clickup/.ignored_contracts/src/rooms.ts
// ============================================================
// WebSocket Room Definitions
// All room names are generated by these functions.
// Agents NEVER construct room name strings manually.
// Emit rules: which events go to which rooms.
// ============================================================

import { RoomPrefix } from './types/enums.js'

// ============================================================
// Room name generators
// ============================================================

export const Rooms = {
  workspace: (workspaceId: string): string =>
    `${RoomPrefix.Workspace}:${workspaceId}`,

  list: (listId: string): string =>
    `${RoomPrefix.List}:${listId}`,

  task: (taskId: string): string =>
    `${RoomPrefix.Task}:${taskId}`,

  user: (userId: string): string =>
    `${RoomPrefix.User}:${userId}`,
} as const

// ============================================================
// Emit rules — which rooms each event type targets
// Services call these to know where to emit
// ============================================================

export const EmitRules = {
  'task.created': (payload: { listId: string; workspaceId: string }) => [
    Rooms.list(payload.listId),
    Rooms.workspace(payload.workspaceId),
  ],

  'task.updated': (payload: { taskId: string; listId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.list(payload.listId),
  ],

  'task.deleted': (payload: { taskId: string; listId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.list(payload.listId),
  ],

  'task.moved': (payload: { taskId: string; oldListId: string; newListId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.list(payload.oldListId),
    Rooms.list(payload.newListId),
  ],

  'task.assigned': (payload: { taskId: string; listId: string; assigneeId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.list(payload.listId),
    Rooms.user(payload.assigneeId),
  ],

  'task.status_changed': (payload: { taskId: string; listId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.list(payload.listId),
  ],

  'comment.created': (payload: { taskId: string; workspaceId: string }) => [
    Rooms.task(payload.taskId),
    Rooms.workspace(payload.workspaceId),
  ],

  'comment.updated': (payload: { taskId: string }) => [
    Rooms.task(payload.taskId),
  ],

  'comment.deleted': (payload: { taskId: string }) => [
    Rooms.task(payload.taskId),
  ],

  'comment.resolved': (payload: { taskId: string }) => [
    Rooms.task(payload.taskId),
  ],

  'doc.created': (payload: { workspaceId: string }) => [
    Rooms.workspace(payload.workspaceId),
  ],

  'doc.updated': (payload: { workspaceId: string }) => [
    Rooms.workspace(payload.workspaceId),
  ],

  'workspace.member_added': (payload: { workspaceId: string; userId: string }) => [
    Rooms.workspace(payload.workspaceId),
    Rooms.user(payload.userId),
  ],

  'workspace.member_removed': (payload: { workspaceId: string; userId: string }) => [
    Rooms.workspace(payload.workspaceId),
    Rooms.user(payload.userId),
  ],

  'notification.send': (payload: { userId: string }) => [
    Rooms.user(payload.userId),
  ],
} as const

// ============================================================
// Subscription rules — what rooms clients join on each view
// Used by the API gateway WebSocket handler
// ============================================================

export interface SubscriptionRules {
  onWorkspaceLoad: (workspaceId: string) => string[]
  onListOpen: (listId: string) => string[]
  onTaskOpen: (taskId: string) => string[]
  onUserConnect: (userId: string) => string[]
}

export const ClientSubscriptions: SubscriptionRules = {
  onWorkspaceLoad: (workspaceId) => [Rooms.workspace(workspaceId)],
  onListOpen: (listId) => [Rooms.list(listId)],
  onTaskOpen: (taskId) => [Rooms.task(taskId)],
  onUserConnect: (userId) => [Rooms.user(userId)],
}
